### PR TITLE
daemon: parse and reject invalid channels in snap ops

### PR DESCRIPTION
**Pull request by chipaca**
This is the last chunk of the first step down this channels path! huzzah.

### CODE REVIEW

**Comment by chipaca** [comment]
@zyga for a user to be bitten by a regression with this they'd have to be poking the REST API by hand, using a bogus channel that just happened to work (e.g. `///stable///`). Because:
1. GUI stores use the channels from the channel map which will be well-formed, and
2. the store would already reject completely broken tracks (e.g. 1/2/3/4), meaning this is just erroring out earlier in this case, and
3. we're cleaning up `///stable///` in `snap` (as well as rejecting completely broken tracks directly here).

File: daemon/api.go [hunk 1]
```diff
@@ -1189,6 +1190,14 @@ func postSnap(c *Command, r *http.Request, user *auth.UserState) Response {
 	}
 	inst.ctx = r.Context()
 
+	if inst.Channel != "" {
+		ch, err := channel.Parse(inst.Channel, "-")
+		if err != nil {
+			return BadRequest("%v", err)
```
Comment (pedronis) [inline daemon/api.go]: do we need/want a special error for ///stable/// ? if not #7670 is probably fine

File: daemon/api.go [hunk 2]
```diff
@@ -1189,6 +1190,14 @@ func postSnap(c *Command, r *http.Request, user *auth.UserState) Response {
 	}
 	inst.ctx = r.Context()
 
+	if inst.Channel != "" {
```
Comment (pedronis) [inline daemon/api.go]: this come from snapRevisionOptions, so we need a check also in api_download.go , maybe we want a helper on snapRevisionOptions ? like validate() ?

File: daemon/api.go (hunk 1 above)
Comment (chipaca) [reply daemon/api.go]: I think it's clear enough as is, and isn't going to be hit by any client we know about

File: daemon/api.go [hunk 3]
```diff
@@ -801,6 +822,30 @@ func (inst *snapInstruction) installFlags() (snapstate.Flags, error) {
 	return flags, nil
 }
 
+func (inst *snapInstruction) validate() error {
```
Comment (mvo5) [inline daemon/api.go]: nice refactor!

**Pull request merged**